crates/vm/src/arch/integration_api.rs

```diff
@@ -304,10 +378,89 @@ pub struct MatrixRecordArena<F> {
     pub trace_offset: usize,
 }
 
-/// An auxiliary struct for implementing [RecordArena] for a single record type.
-/// This is useful for chips that have a single row per instruction,
-/// and where we only want to allocate a row with compressed information,
-/// not a trace row to be filled later.
+impl<F: Field> MatrixRecordArena<F> {
+    pub fn alloc_single_row(&mut self) -> &mut [u8] {
+        self.alloc_buffer(1)
+    }
+
+    pub fn alloc_buffer(&mut self, num_rows: usize) -> &mut [u8] {
+        let start = self.trace_offset;
+        self.trace_offset += num_rows * self.width;
+        let row_slice = &mut self.trace_buffer[start..self.trace_offset];
+        let size = size_of_val(row_slice);
+        let ptr = row_slice as *mut [F] as *mut u8;
+        // SAFETY:
+        // - `ptr` is non-null
+        // - `size` is correct
+        // - alignment of `u8` is always satisfied
+        unsafe { &mut *std::ptr::slice_from_raw_parts_mut(ptr, size) }
+    }
+}
+
+impl<F: Field> RowMajorMatrixArena<F> for MatrixRecordArena<F> {
+    fn set_capacity(&mut self, trace_height: usize) {
+        let size = trace_height * self.width;
+        // PERF: use memset
+        self.trace_buffer.resize(size, F::ZERO);
+    }
+
+    fn with_capacity(height: usize, width: usize) -> Self {
+        let trace_buffer = F::zero_vec(height * width);
+        Self {
+            trace_buffer,
+            width,
+            trace_offset: 0,
+        }
+    }
+
+    fn width(&self) -> usize {
+        self.width
+    }
+
+    fn trace_offset(&self) -> usize {
+        self.trace_offset
+    }
+
+    fn into_matrix(self) -> RowMajorMatrix<F> {
+        RowMajorMatrix::new(self.trace_buffer, self.width)
+    }
+}
+
+/// [RecordArena] implementation for [MatrixRecordArena], with [AdapterCoreLayout]
+/// **NOTE**: `A` is the adapter RecordMut type and `C` is the core RecordMut type
+impl<'a, F: Field, A, C, M: AdapterCoreMetadata> RecordArena<'a, AdapterCoreLayout<M>, (A, C)>
+    for MatrixRecordArena<F>
+where
+    [u8]: CustomBorrow<'a, A, AdapterCoreLayout<M>> + CustomBorrow<'a, C, AdapterCoreLayout<M>>,
+    M: Clone,
+{
+    fn alloc(&'a mut self, layout: AdapterCoreLayout<M>) -> (A, C) {
+        let adapter_width = M::get_adapter_width();
```

Comment:
I guess this happens to be ok for our use cases now, but just noting it would not support the case if the adapter width was stored in the adapter step struct and dynamically determined instead